ROCm Sparsity through HipSparseLT

petrex opened the pull request:
TLDR:  

- This pull request introduces support for hipSPARSELt in ROCm, current usage would be semi-structure sparsity. 
- Require **ROCm 6.4** && **gfx942/gfx950**. 
- The average performance uplift (compare to dense operation) is ~ 20% in ROCm 6.4 but expect further performance lift along the way.

          
### Dense vs. Sparse Performance Comparison

#### **NT (Row-major)**  
**Average Uplift**: `1.20`

| M     | N      | K      | hipsparselt-bench (us) | hipblaslt-bench get all (us) | Uplift |
|-------|--------|--------|-------------------------|-------------------------------|--------|
| 14336 | 8      | 4096   | 20.05                   | 25.3                          | 1.26   |
| 4096  | 8      | 14336  | 21.07                   | 25.28                         | 1.20   |
| 3072  | 3072   | 10240  | 299.05                  | 351.82                        | 1.18   |
| 3072  | 1536   | 768    | 18.56                   | 20.05                         | 1.08   |
| 3072  | 17664  | 768    | 163.13                  | 173.91                        | 1.07   |
| 3072  | 196608 | 768    | 1717.30                 | 1949.63                       | 1.14   |
| 3072  | 24576  | 768    | 206.84                  | 242.98                        | 1.17   |
| 3072  | 6144   | 768    | 53.90                   | 56.88                         | 1.06   |
| 3072  | 98304  | 768    | 833.77                  | 962.28                        | 1.15   |
| 768   | 1536   | 768    | 8.53                    | 19.65                         | 2.30   |
| 768   | 17664  | 768    | 46.02                   | 46.84                         | 1.02   |
| 768   | 196608 | 768    | 463.15                  | 540.46                        | 1.17   |
| 768   | 24576  | 768    | 54.32                   | 59.55                         | 1.10   |
| 768   | 6144   | 768    | 19.47                   | 20.15                         | 1.03   |
| 768   | 98304  | 768    | 231.88                  | 258.73                        | 1.12   |

---

#### **NN (Row-major)**  
**Average Uplift**: `1.13`

| M   | N      | K     | hipsparselt-bench (us) | hipblaslt-bench get all (us) | Uplift |
|-----|--------|-------|-------------------------|-------------------------------|--------|
| 768 | 1536   | 3072  | 27.50                   | 28.78                         | 1.05   |
| 768 | 17664  | 3072  | 125.06                  | 158.94                        | 1.27   |
| 768 | 196608 | 3072  | 1568.38                 | 1767.12                       | 1.13   |
| 768 | 24576  | 3072  | 171.05                  | 203.49                        | 1.19   |
| 768 | 6144   | 3072  | 58.72                   | 60.39                         | 1.03   |
| 768 | 98304  | 3072  | 787.15                  | 887.60                        | 1.13   |

-------------------------


This pull request introduces support for hipSPARSELt in ROCm, alongside various updates and improvements to the codebase and test suite. The changes primarily involve adding configuration flags, updating conditional checks, and ensuring compatibility with hipSPARSELt.

### ROCm and hipSPARSELt Support:

* [`BUILD.bazel`](diffhunk://#diff-7fc57714ef13c3325ce2a1130202edced92fcccc0c6db34a72f7b57f60d552a3R292): Added `@AT_HIPSPARSELT_ENABLED@` substitution to enable hipSPARSELt support.
* [`aten/CMakeLists.txt`](diffhunk://#diff-0604597797bb21d7c39150f9429d6b2ace10b79ab308514ad03f76153ae8249bR104-R110): Introduced a conditional flag to enable hipSPARSELt support based on ROCm version.
* [`aten/src/ATen/CMakeLists.txt`](diffhunk://#diff-ce80f3115ab2f6be5142f0678a1fc92c6b2d7727766ce44f48726c99e720f777R37): Added `AT_HIPSPARSELT_ENABLED` configuration.
* [`aten/src/ATen/cuda/CUDAConfig.h.in`](diffhunk://#diff-8bb82da825ca87c28233abacffa1b0566c73a54990b7a77f3f5108d3718fea15R11): Defined `AT_HIPSPARSELT_ENABLED` macro.
* `caffe2/CMakeLists.txt`, `cmake/Dependencies.cmake`, `cmake/public/LoadHIP.cmake`: Included hipSPARSELt in the ROCm dependencies. [[1]](diffhunk://#diff-c5ee05f1e918772792ff6f2a3f579fc2f182e57b1709fd786ef6dc711fd68b27R1380) [[2]](diffhunk://#diff-12e8125164bbfc7556b1781a8ed516e333cc0bf058acb7197f7415be44606c72L1084-R1084) [[3]](diffhunk://#diff-b98e27b9a5f196a6965a99ee5a7bb15b3fc633d6375b767635b1b04ccb2fd3d5R153)

### Codebase Updates:

* [`aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp`](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R1-R6): Added hipSPARSELt support checks and initialization functions. Updated various methods to conditionally handle hipSPARSELt. [[1]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R1-R6) [[2]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R22-R67) [[3]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R78-R85) [[4]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R97-R109) [[5]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R183-R188) [[6]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3L134-R200) [[7]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3R213-R222) [[8]](diffhunk://#diff-ae921dd1584ab98fdd9c25a3521047795de702223f5b65fdaa45a5bd92b4d1f3L217-R285)

### Test Suite Updates:

* [`test/test_sparse_semi_structured.py`](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR50-R65): Added checks for hipSPARSELt availability and updated test conditions to skip tests not supported on ROCm. [[1]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR50-R65) [[2]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR228) [[3]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR239) [[4]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR250) [[5]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR579) [[6]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR624) [[7]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR661) [[8]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR695) [[9]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR730) [[10]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR755) [[11]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR771) [[12]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR809) [[13]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR844) [[14]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cL840-R854) [[15]](diffhunk://#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR1005)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

Commit:
Add hipSPARSELt support for ROCm in CMake and CUDA files

- Introduced a new flag for hipSPARSELt support in CMakeLists.txt, enabling it for ROCM versions 6.4.0 and above.
- Updated CUDAConfig.h.in to define the AT_HIPSPARSELT_ENABLED macro.
- Enhanced cuSPARSELtOps.cpp with initialization logic for hipSparseLt support, including architecture checks and error handling.

This change improves compatibility with ROCm's hipSparseLt library, allowing for better performance on supported architectures.

Commit:
Enhance ROCm support by adding hipSPARSELt dependencies in CMake and CUDA mappings

- Updated Dependencies.cmake to include hipSPARSELt in the list of public HIP dependency libraries.
- Modified cuda_to_hip_mappings.py to map cuSPARSELt headers and types to their hipSPARSELt counterparts, improving compatibility and functionality for ROCm users.

These changes facilitate better integration with ROCm's hipSPARSELt library, enhancing performance and usability.

Commit:
Enhance ROCm support in cuSPARSELt operations and CMake configuration

- Added checks for hipSparseLt support in cuSPARSELt operations, ensuring compatibility with specific ROCm architectures and versions.
- Updated CMakeLists.txt to include the hipSPARSELt include directory, improving build configuration for ROCm users.
- Enhanced LoadHIP.cmake to find and include the hipSPARSELt package, further integrating ROCm support.

These changes improve the functionality and performance of sparse operations on ROCm platforms.

Commit:
Update error messages and conditional compilation for cuSPARSELt operations

- Modified error message for unsupported data types in cuSPARSELt to include hipSparseLt for clarity.
- Enhanced conditional compilation checks for Float8 support to exclude ROCm builds, ensuring compatibility with specific CUDA versions.

These changes improve error handling and maintain compatibility across different GPU architectures.

Commit:
Add hipSPARSELt support in BUILD.bazel and CMakeLists.txt

- Introduced a new substitution for AT_HIPSPARSELT_ENABLED in BUILD.bazel to enable hipSPARSELt support.
- Updated CMakeLists.txt to set the AT_HIPSPARSELT_ENABLED flag based on CAFFE2_USE_HIPSPARSELT, enhancing build configuration for ROCm users.

These changes improve the integration of hipSPARSELt support, ensuring better compatibility and performance on ROCm platforms.

Commit:
Enhance ROCm support in sparse semi-structured tests

- Added checks for hipSPARSELt availability in the test suite, improving compatibility with ROCm platforms.
- Updated inference and training data types to conditionally include float32 for non-HIP environments.
- Introduced skip conditions for ROCm in various test cases to prevent execution on unsupported platforms.

These changes ensure better integration and testing for ROCm users, enhancing the robustness of sparse operations.

pruthvistony commented:
moving this PR to draft till ROCm 6.4 is released and CI is updated.

Commit:
Refine supported architectures for hipSparseLt in cuSPARSELt operations

- Removed unsupported architectures (gfx1200, gfx1201) from the list of supported architectures for hipSparseLt.
- Updated error messages to reflect the current supported architectures, ensuring clarity for users regarding compatibility.

These changes enhance the accuracy of architecture support for hipSparseLt, improving user experience and error handling.

Commit:
Refactor hipSparseLt support initialization in cuSPARSELt operations

- Simplified conditional compilation for hipSparseLt support by removing redundant checks.
- Cleaned up whitespace in the initialization function for better readability.

These changes enhance the clarity and maintainability of the code related to hipSparseLt support, ensuring a more streamlined implementation.

Commit:
Refactor hipSparseLt initialization in cuSPARSELt operations

- Changed the initialization flag from std::once_flag to c10::once_flag for better integration with the C10 library.
- Updated the call to initialize hipSparseLt support to use c10::call_once, ensuring consistency in the codebase.

These changes improve the maintainability and compatibility of hipSparseLt support in the cuSPARSELt operations.

Commit:
Update hipSparseLt initialization to use c10::call_once in cuSPARSELt operations

- Changed the initialization call from std::call_once to c10::call_once for improved integration with the C10 library.
- This update enhances consistency and maintainability in the hipSparseLt support implementation.

These changes contribute to a more streamlined and cohesive codebase.

petrex commented:
FYI @jcaip

Commit:
Refactor hipSparseLt support initialization in cuSPARSELtOps.cpp

Removed the architecture check loop and simplified the initialization logic to directly check for GPU architecture support using the CUDA hooks. This change enhances performance by reducing unnecessary iterations and focuses on the first available device only.

Commit:
Remove unused includes in cuSPARSELtOps.cpp to streamline code and improve readability.

Commit:
Refactor hipSparseLt support check in cuSPARSELtOps.cpp

Updated the isHipSparseLtSupported function to remove the device index parameter, simplifying the support check logic. This change enhances code clarity and maintains the integrity of the architecture support validation for hipSparseLt.

Commit:
Add ROCM support check in cuSPARSELtOps.cpp

Introduced a conditional compilation check for hipSparseLt support on ROCM devices. This ensures that the code only executes the support validation when compiled with ROCM, enhancing compatibility and maintainability of the sparse matrix operations.

jeffdaily reviewed (approved):
Approved, pending clean CI.

petrex commented:
@pytorchbot rebase

petrex commented on aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp:
```diff
@@ -120,6 +164,14 @@ std::tuple<at::Tensor, int64_t, int64_t, int64_t, int64_t> _cslt_sparse_mm_impl(
   cusparseComputeType compute_type;
   auto compression_factor = 9;
 
+  #ifdef USE_ROCM
+  TORCH_CHECK(
+      isHipSparseLtSupported(),
+      "hipSparseLt not supported on this device, supported architectures: "
+      "gfx950, gfx942."
+      "required ROCM version: 6.4.0 or later.");
+  #endif
+
```
No problem. were you suggesting add the msgs inside `isHipSparseLtSupported()`?

petrex replied:
just  pushed

Commit:
Enhance hipSparseLt support validation in cuSPARSELtOps.cpp

Added a direct check for hipSparseLt support in the isHipSparseLtSupported function, ensuring that an explicit error message is provided when the feature is not supported on the device. This improves error handling and user feedback regarding ROCM compatibility.

Commit:
Remove unused ROCM include directories from CMakeLists.txt to streamline build configuration and improve maintainability.

Commit:
Refactor inference data types in test_sparse_semi_structured.py

Removed conditional assignment for inference_dtypes to simplify the code. This change enhances clarity and maintains consistency in the data type definitions used for testing sparse semi-structured tensors.

jeffdaily commented:
@pytorchbot merge
